src/data/arena-tag.ts

```diff
@@ -913,6 +913,121 @@ export class DelayedAttackTag extends ArenaTag {
   }
 }
 
+interface PendingHealEffect {
+  readonly sourceId: number;
+  readonly moveId: MoveId;
+  readonly restorePP: boolean;
+  readonly healMessageKey: string;
+}
+
+/**
+ * Arena tag to contain stored healing effects, namely from
+ * {@link https://bulbapedia.bulbagarden.net/wiki/Healing_Wish_(move) | Healing Wish}
+ * and {@link https://bulbapedia.bulbagarden.net/wiki/Lunar_Dance_(move) | Lunar Dance}.
+ * When a damaged Pokemon first enters the effect's {@linkcode BattlerIndex | field position},
+ * their HP is fully restored, and they are cured of any non-volatile status condition.
+ * If the effect is from Lunar Dance, their PP is also restored.
+ * @extends ArenaTag
+ */
+export class PendingHealTag extends ArenaTag {
+  private pendingHeals: Map<BattlerIndex, PendingHealEffect[]>;
+
+  constructor() {
+    super(ArenaTagType.PENDING_HEAL, 0);
+    this.pendingHeals = new Map<BattlerIndex, PendingHealEffect[]>();
+  }
+
+  /**
+   * Adds a pending healing effect to the field. Effects under the same move *and*
+   * target index as an existing effect are ignored.
+   * @param targetIndex - The {@linkcode BattlerIndex} under which the effect applies
+   * @param healEffect - The {@linkcode PendingHealEffect | data} for the pending heal effect
+   */
+  public addHeal(targetIndex: BattlerIndex, healEffect: PendingHealEffect): void {
+    if (this.pendingHeals.has(targetIndex)) {
+      const existingHealEffects = this.pendingHeals.get(targetIndex);
+      if (existingHealEffects?.every((he) => he.moveId !== healEffect.moveId)) {
```

Comment:
NIT: Wouldn't checking the reverse (`!...some()...`) be faster? That way it doesn't have to go through the whole list.